pyroomacoustics/bss/ilrma.py

```diff
@@ -0,0 +1,146 @@
+'''
+Blind Source Separation using Independent Low-Rank Matrix Analysis (ILRMA)
+
+2018 (c) Juan Azcarreta, Robin Scheibler, MIT License
+'''
+import numpy as np
+from .common import projection_back
+
+def ilrma(X, n_src=None, n_iter=20, proj_back=False, W0=None,
+        n_components=2,
+        return_filters=0,
+        callback=None):
+
+    '''
+    Implementation of ILRMA algorithm without partitioning function for BSS presented in
+
+    D. Kitamura, N. Ono, H. Sawada, H. Kameoka, H. Saruwatari, *Determined blind
+    source separation unifying independent vector analysis and nonnegative matrix
+    factorization,* IEEE/ACM Trans. ASLP, vol. 24, no. 9, pp. 1626-1641, September 2016
+
+    D. Kitamura, N. Ono, H. Sawada, H. Kameoka, and H. Saruwatari *Determined Blind Source Separation
+    with Independent Low-Rank Matrix Analysis*, in Audio Source Separation, S. Makino, Ed. Springer, 2018, pp.  125-156.
+
+    Parameters
+    ----------
+    X: ndarray (nframes, nfrequencies, nchannels)
+        STFT representation of the observed signal n_src: int, optional
+        The number of sources or independent components
+    n_iter: int, optional
+        The number of iterations (default 20)
+    proj_back: bool, optional
+        Scaling on first mic by back projection (default True)
+    W0: ndarray (nfrequencies, nchannels, nchannels), optional
+        Initial value for demixing matrix
+    n_components: int
+        Number of components in the non-negative spectrum
+    return_filters: bool
+        If true, the function will return the demixing matrix too
+    callback: func
+        A callback function called every 10 iterations, allows to monitor convergence
+
+    Returns
+    -------
+    Returns an (nframes, nfrequencies, nsources) array. Also returns
+    the demixing matrix W (nfrequencies, nchannels, nsources)
+    if ``return_values`` keyword is True.
+    '''
+    X = np.transpose(X, (1, 0, 2))  # now X is (nfrequencies, nframes, nchannels)
+    n_freq, n_frames, n_chan = X.shape
+
+    # default to determined case
+    if n_src is None:
+        n_src = X.shape[2]
+
+    # Only supports determined case
+    assert n_chan == n_src
+
+    # initialize the demixing matrices
+    # The demixing matrix has the following dimensions (nfrequencies, nchannels, nsources),
+    if W0 is None:
+        W = np.array([np.eye(n_chan, n_src) for f in range(n_freq)], dtype=X.dtype)
+    else:
+        W = W0.copy()
+
+    # initialize the nonnegative matrixes with random values
+    T = abs(np.array(np.random.rand(n_freq, n_components, n_src)))
```

Comment:
I think it should be possible to initialize the `T` and `V` matrices so that they have a power profile matching the input spectrogram. This might give faster convergence of the algorithm.
```
S  # input power spectrogram
P_freq = S.mean(axis=0) / n_src
P_time = S.mean(axis=1) / n_src

T_init = np.random.rand(n_freq, n_components, n_src) * P_freq[:,None,None]
V_init = np.random.rand(n_components, n_frames, n_src) * P_time[None,:,None]
```
Also, I think `rand` gives a value in `[0,1]`. Why is `abs` needed here ?